builder/dockerfile/internals.go

```diff
@@ -569,8 +575,13 @@ func (b *Builder) run(cID string) (err error) {
 			Code:    ret,
 		}
 	}
-
-	return nil
+	once.Do(finish)
```

Comment:
This seems reasonable, but I think you should have an error channel that the goroutine sends `errCancelled` to if it's cancelled, and you should read from this channel before the `select` below. Otherwise the `select` can still happen before the goroutine finishes.